vyper/venom/analysis.py

```diff
@@ -60,29 +62,31 @@ def _calculate_liveness_bb(bb: IRBasicBlock) -> None:
             liveness.remove(out)
         instruction.liveness = liveness
 
+    return orig_liveness != bb.instructions[0].liveness
 
-def _calculate_liveness_r(bb: IRBasicBlock, visited: dict) -> None:
-    assert isinstance(visited, dict)
-    for out_bb in bb.cfg_out:
-        if visited.get(bb) == out_bb:
-            continue
-        visited[bb] = out_bb
-
-        # recurse
-        _calculate_liveness_r(out_bb, visited)
 
+def _calculate_out_vars(bb: IRBasicBlock) -> bool:
+    """
+    Compute out_vars of basic block.
+    Returns True if out_vars changed
+    """
+    out_vars = bb.out_vars.copy()
+    for out_bb in bb.cfg_out:
         target_vars = input_vars_from(bb, out_bb)
-
-        # the output stack layout for bb. it produces a stack layout
-        # which works for all possible cfg_outs from the bb.
         bb.out_vars = bb.out_vars.union(target_vars)
-
-    _calculate_liveness_bb(bb)
+    return out_vars != bb.out_vars
 
 
 def calculate_liveness(ctx: IRFunction) -> None:
     _reset_liveness(ctx)
-    _calculate_liveness_r(ctx.basic_blocks[0], dict())
+    while True:
+        changed = False
+        for bb in ctx.basic_blocks:
+            changed |= _calculate_out_vars(bb)
+            changed |= _calculate_liveness(bb)
```

Comment:
hmm this seems like maybe a performance problem (but i don't have a solution in mind)